openhands/controller/stuck.py

```diff
@@ -37,31 +39,38 @@ def is_stuck(self):
             )
         ]
 
-        # it takes 3 actions minimum to detect a loop, otherwise nothing to do here
-        if len(filtered_history) < 3:
+        # It takes 4 actions minimum to detect a loop, otherwise nothing to do here
+        # Sonnet often changes approach after 3 errors on its own, so require 4!
```

Comment:
Sonnet 3.5 is very very good with stuck loops. Somehow Anthropic basically solved the problem, it seems. We have seen a run of SWE-bench lite where it had 3/300, while gpt-4o had usually like 60/300 or so. Also, I've been encountering other scenarios outside OH (non-code agents) where Claude jumps out of loops if it gets near at all, while GPTs, Llama 3.1 and others keep getting stuck.

Very interesting that it got near and got out! When Sonnet changed approach, did it succeed or did it get stuck with 4x later? What specific case was it, from the four cases of loops that we check for? Have we seen Sonnet fall into all of these four possible loops?

We had a different number of iterations for different cases, and that makes sense I think, they don't have to be the same. The behavior depends on the agent and llm. The risk of making them all 4 is that the LLM can change a too slight bit, and then we fail to detect it for a long time.

Reply:
It's okay, I ask because it's possible we'd be better off changing from 3 to 4 on _that_ scenario, but we don't have to change all scenarios. The (possible) bad side effect can be seen in that test I mentioned below, reproducing system prompt: if requiring 4 identical/similar steps results in 20-40 useless tries until we stop the loop, then... we should use 3.

Unfortunately that one is only good for the syntax error scenario, and I don't have good "live" tests for the other scenarios. Just using the agent a lot can show us others...

Reply:
I checked the log file and I see the agent got stuck with `_is_stuck_repeating_action_observation` check, not the syntax error loop check. That's a little unfortunate I guess? But this PR doesn't address that. So it's not clear to me we're gaining anything by trying to change some 3 to 4.